sympy/core/kind.py

```diff
@@ -177,4 +184,196 @@ def __new__(cls):
     def __repr__(self):
         return "BooleanKind"
 
-BooleanKind = BooleanKind()
+BooleanKind = _BooleanKind()
+
+
+class KindDispatcher:
```

Comment:
This reminds me on the doc about [type promotion](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/) in the Julia programming language.

Reply:
Maybe better rename this class `KindPromoter`? It's not really a dispatcher in the end.

Reply:
I don't know Julia well, but it seem that promoting is converting the arguments with different types to one single type. `KindDispatcher`, on the other hand, does not change the argument type so I'm -1 to this naming.